### DISCUSSION TITLE
Having trouble installing from source code

### DISCUSSION THREAD

**Discussion by drjswalker** [FrancescoCeruti/linux-show-player#291] [Q&A]
I unpacked the source code  (master) into a directory and am at a loss as to what to do mext. I ran python main.py but it complained about the lisp module being missing. Any ideas what to do next?

**Comment by FrancescoCeruti** [comment]
Take a look at these instructions: https://linux-show-player-users.readthedocs.io/en/develop/installation.html#manually

**Reply by kzembower** [reply]
I would like to suggest that the instructions under "Manually" could be improved. 

I couldn't tell which of the 'Distributions dependencies' was to be installed via the package manager, or with python pip. For instance, on my Ubuntu 22.04 system, the package managers is unable to locate package 'gstreamer.' Also, there's no 'python' but it suggests: "2to3 python2-minimal:i386 python2:i386 python2-minimal python2 dh-python
  python-is-python3". I chose 'python-is-python3'. And again, when it mentions 'gstreamer-plugins-(good/ugly/bad)', does this mean install three packages? From where (Ubuntu 22.04 can't find any of the three)?

Dispite LiSP 0.6.0.dev0 seeming to run and perform normally (can create file, create audio cue, play it, etc.), I get these errors when I run it:
```
$ poetry run linux-show-player
Warning: 'linux-show-player' is an entry point defined in pyproject.toml, but it's not installed as a script. You may get improper `sys.argv[0]`.

The support to run uninstalled scripts will be removed in a future release.

Run `poetry install` to resolve and get rid of this message.

2023-11-25 15:23:54.918	lisp.core.loading	WARNING	Cannot load python module: "alsa_sink"
Traceback (most recent call last):
  File "/home/kevin/linux-show-player/lisp/core/loading.py", line 65, in load_modules
    yield mod_name, import_module(mod_path)
  File "/home/kevin/linux-show-player/lisp/core/loading.py", line 166, in import_module
    return __import__(module_path, globals(), locals(), ["*"])
  File "/home/kevin/linux-show-player/lisp/plugins/gst_backend/config/alsa_sink.py", line 21, in <module>
    from lisp.plugins.gst_backend.settings.alsa_sink import AlsaSinkSettings
  File "/home/kevin/linux-show-player/lisp/plugins/gst_backend/settings/alsa_sink.py", line 25, in <module>
    from pyalsa import alsacard
ModuleNotFoundError: No module named 'pyalsa'
2023-11-25 15:23:54.930	lisp.core.loading	WARNING	Cannot load python module: "alsa_sink"
Traceback (most recent call last):
  File "/home/kevin/linux-show-player/lisp/core/loading.py", line 65, in load_modules
    yield mod_name, import_module(mod_path)
  File "/home/kevin/linux-show-player/lisp/core/loading.py", line 166, in import_module
    return __import__(module_path, globals(), locals(), ["*"])
  File "/home/kevin/linux-show-player/lisp/plugins/gst_backend/settings/alsa_sink.py", line 25, in <module>
    from pyalsa import alsacard
ModuleNotFoundError: No module named 'pyalsa'
2023-11-25 15:23:54.967	lisp.core.loading	WARNING	Cannot load python module: "artnet"
Traceback (most recent call last):
  File "/home/kevin/linux-show-player/lisp/core/loading.py", line 65, in load_modules
    yield mod_name, import_module(mod_path)
  File "/home/kevin/linux-show-player/lisp/core/loading.py", line 166, in import_module
    return __import__(module_path, globals(), locals(), ["*"])
  File "/home/kevin/linux-show-player/lisp/plugins/timecode/protocols/artnet.py", line 21, in <module>
    from ola.OlaClient import OlaClient, OLADNotRunningException
ModuleNotFoundError: No module named 'ola'
$ 
```

Any suggestions how to fix these errors? I've run 'poetry install' and gotten:
```
$ poetry install
Installing dependencies from lock file

Package operations: 1 install, 1 update, 0 removals

  • Downgrading idna (3.6 -> 3.4)
  • Installing pyalsa (1.2.7 7d6dfe0): Failed

  ChefBuildError

  Backend subprocess exited when trying to invoke build_wheel
  
  /tmp/tmph272m9x1/.venv/lib/python3.10/site-packages/setuptools/_distutils/dist.py:250: UserWarning: 'licence' distribution option is deprecated; use 'license'
    warnings.warn(msg)
  running bdist_wheel
  running build
  running build_py
  copying pyalsa/__init__.py -> build/lib.linux-x86_64-cpython-310/pyalsa
  running build_ext
  building 'pyalsa.alsacard' extension
  x86_64-linux-gnu-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/tmp/tmph272m9x1/.venv/include -I/usr/include/python3.10 -c pyalsa/alsacard.c -o build/temp.linux-x86_64-cpython-310/pyalsa/alsacard.o
  pyalsa/alsacard.c:25:10: fatal error: alsa/asoundlib.h: No such file or directory
     25 | #include "alsa/asoundlib.h"
        |          ^~~~~~~~~~~~~~~~~~
  compilation terminated.
  error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
  

  at ~/.local/pipx/venvs/poetry/lib/python3.10/site-packages/poetry/installation/chef.py:164 in _prepare
      160│ 
      161│                 error = ChefBuildError("\n\n".join(message_parts))
      162│ 
      163│             if error is not None:
    → 164│                 raise error from None
      165│ 
      166│             return path
      167│ 
      168│     def _prepare_sdist(self, archive: Path, destination: Path | None = None) -> Path:
$ 
```

Thanks for any thoughts or advice.

-Kevin

**Reply by s0600204** [reply]
*Disclaimer*: I think I've got package names correct below, but as I don't have access to a "live" Ubuntu/Debian/Fedora/Gentoo/CentOS system, I haven't tested to make sure.

To quote the guide:

> You should install these libraries/packages using your package manager

None of the list needs installing via `pip` - everything that could be installed from the PyPI is done so via `poetry`. (Except `poetry` itself, as we need `poetry` installed before running `poetry`.)

Linux Show Player requires `python` `3.8` or better. Most Linux distributions have transitioned to `python` `3.x` being the default (over `python` `2.x`)  (citation needed), and some have even removed `python` `2.x` entirely. However for those distributions (os OSes, such as macOS) where such things matter, `python3` should be substituted where necessary. (If you've previously been running Linux Show Player 0.5 successfully, then you should already be okay in this regard.)

You appear to have `poetry` already, but for completeness sake:

* `python3-poetry` (Debian, Ubuntu, other Debian-derived distros)
* `python-poetry` (ArchLinux)
* `poetry` (Gentoo, Fedora)

----

GStreamer works on the basis of plugins - small(ish) libraries that each handle a specific codec, file type, audio-api, etc. Most of these plugins are placed into one of four groupings: Base, Good, Ugly, and Bad. The reasoning for this can be found here: https://gstreamer.freedesktop.org/documentation/additional/splitup.html.

For a *functional* system, only the Base plugins need to be installed. For an actually *useful* system, one or all three of the others should also be installed, depending on which file types you expect to encounter, your taste for well constructed/maintained/tested code, and how well you wish to adhere to the ideals of free-as-in-freedom versus free-as-in-free-beer.

There are some plugins that don't get bundled into these, and must be installed separately. `gst-libav` is one of them.

On a Debian, Ubuntu, and such-derived systems, these can be found as:

* `gstreamer1.0-plugins-good`
* `gstreamer1.0-plugins-ugly`
* `gstreamer1.0-plugins-bad`
* `gstreamer1.0-libav`

For ArchLinux, Gentoo systems:

* `gst-plugins-good`
* `gst-plugins-ugly`
* `gst-plugins-bad`
* `gst-libav` (ArchLinux)
* `gst-plugins-libav` (Gentoo)

For Fedora, CentOS systems:

* `gstreamer1-plugins-good`
* `gstreamer1-plugins-ugly-free`
* `gstreamer1-plugins-bad-free`
* `gstreamer1-plugin-libav` (Not available on CentOS)

(Installing any one of the above should also install the Base plugins as a dependency, which itself pulls in GStreamer.)

----

From the `poetry install` failure you appear to have not installed the `alsa-lib` development files:

* `libasound2-dev` (Debian, Ubuntu, and other derived)
* `alsa-lib-devel` (Fedora, CentOS)

(ArchLinux and Gentoo users should have them installed so long as they have ALSA itself installed.)

----

From the `poetry run linux-show-player` errors, you appear to have not installed the Open Lighting Architecture ("OLA"). This is only used (at the moment) to add support for a particular form of timecode, so you can safely ignore this. Also, the installation guide makes the following note:

> You might also need to install ola, for ArtNet timecode. However, using the method described below, it’s not possile to use it.

Not sure why not, but I'm sure Francesco will elaborate. 

If you *do* wish to install it:

* `ola`

Which is also the package name for users on Gentoo, as well as those on ArchLinux (albeit available from the AUR). Fedora and CentOS distros don't appear to carry it.

----

The development files for `rtmidi` shouldn't, I think, be needed - `poetry` will either decide to use the `manylinux` wheel from PyPI - which contains a prebuilt connection to `rtmidi` inside - or it will download and build from the tarball, also from PyPI - which contains the development files.

----

The development files for `liblo` should be:

* `liblo-dev` (Debian and derived)
* `liblo-devel` (Fedora)

(And once again, ArchLinux and Gentoo can just install `liblo`. And CentOS doesn't carry it.)

----

What's not mentioned in the build instructions is the requirement for Qt5 and Qt5-SVG:

* `libqt5svg5` (Debian, Ubuntu, etc.)
* `qt5-svg` (ArchLinux)
* `qt5-qtsvg` (Fedora, CentOS)
* `qtsvg` (Gentoo - appears to be the Qt6 version, but their distro-provided version of LiSP 0.5.3 indirectly depends on it, so it must work? :shrug:)

Any of which should also pull in Qt5 as a dependency.

----

Think that's the bulk of it. Sorry for the wall of text.

**Reply by FrancescoCeruti** [reply]
Thanks @s0600204 :+1: I'll update the docs to include distro-specific package name.

> > You might also need to install ola, for ArtNet timecode. However, using the method described below, it’s not possile to use it.
>
> Not sure why not, but I'm sure Francesco will elaborate.

To use `ola` we need two pieces, the daemon and the python client library. AFAIK we're missing an "official" pypi package, https://github.com/OpenLightingProject/ola/issues/1451. Also I'm not sure how it might behave if different versions of daemon and client are installed.

> What's not mentioned in the build instructions is the requirement for Qt5 and Qt5-SVG

The qt binaries are provided via pypi by [`pyqt5-qt5`](https://pypi.org/project/PyQt5-Qt5/) (they already ship the SVG plugin)

**Reply by kzembower** [reply]
@s0600204 and @FrancescoCeruti , thank you both so much. No apology necessary for all the text; it was just what I needed.

I apologize in advance for asking you to troubleshoot my individual problems, but I'm stuck again. I verified and installed all the software that @s0600204 mentioned, but now have an error that prevents LiSP from running at all.

When I was trying on my own to set up LiSP, I installed poetry via pipx. Then, following @s0600204's directions, I installed it again with apt-get install python3-poetry. When I tried to run LiSP, I got:
```
$ poetry run linux-show-player

  RuntimeError

  The Poetry configuration is invalid:
    - Additional properties are not allowed ('group' was unexpected)
  

  at /usr/lib/python3/dist-packages/poetry/core/factory.py:43 in create_poetry
       39│             message = ""
       40│             for error in check_result["errors"]:
       41│                 message += "  - {}\n".format(error)
       42│ 
    →  43│             raise RuntimeError("The Poetry configuration is invalid:\n" + message)
       44│ 
       45│         # Load package
       46│         name = local_config["name"]
       47│         version = local_config["version"]
$
```
Trying to fix this, I removed poetry with 'pip uninstall poetry' which seemed to work. However, I still get the error. In addition, the file ~/.config/pypoetry/config.toml is present, but with zero size. I've also tried running 'poetry config' but it seems to do nothing, and the size of config.toml is unchanged. 'poetry' points to:
```
$ which poetry
/usr/bin/poetry
$ ll /usr/bin/poetry 
-rwxr-xr-x 1 root root 213 Jan 11  2023 /usr/bin/poetry*
$ 
```

Any suggestions on what to do next? Thank you for any advice or directions you can offer.

-Kevin

**Reply by FrancescoCeruti** [reply]
No problem :-) The version from the repository is likely too old. Remove that version, `apt remove python3-poetry`, and reinstall with pipx, it should work.

**Reply by s0600204** [reply]
> The version from the repository is likely too old

`1.1.12` according to [Repology](https://repology.org/project/poetry/versions), so yes it is. My mistake not double checking the version when collating package names above. (Users of Fedora 37 and earlier would also need to install `poetry` from the PyPI.)
